### PR TITLE
add Tikhonov regularization to optimization reporter

### LINKED ISSUE #24101: Implement Tikhonov regularization
## Reason

To help with convergence, especially with underdetermined systems.
## Design

Add the Tikhonov regularization term to objective function, gradient and hessian.
## Impact

Helps with convergence

### PR DESCRIPTION
**Pull request by lynnmunday**
closes #24101

### CODE REVIEW

**Comment by lynnmunday** [comment]
@zachmprince will you review and merge if this is good?

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 1]
```diff
@@ -71,15 +81,22 @@ OptimizationReporterBase::setSimulationValuesForTesting(std::vector<Real> & data
...
-                 _nvalues[p],
+                 _nvalues[param_group_id],
                  " versus ",
-                 _gradients[p]->size(),
+                 _gradients[param_group_id]->size(),
                  ".");
+    auto params = _parameters[param_group_id];
+    auto grads = _gradients[param_group_id];
+    for (const auto & param_id : make_range(_nvalues[param_group_id]))
+      grads->at(param_id) += params->at(param_id) * _tikhonov_coeff;
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:97]:
```suggestion
      (*grads)[param_id] += (*params)[param_id] * _tikhonov_coeff;
```

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 2]
```diff
@@ -51,6 +54,13 @@ OptimizationReporterBase::computeObjective()
   for (auto & misfit : _misfit_values)
     val += misfit * misfit;
 
+  Real param_norm_sqr = 0;
+  for (const auto & data : _parameters)
+    for (const auto & val : *data)
+      param_norm_sqr += val * val;
+
+  val += _tikhonov_coeff * param_norm_sqr;
```
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:62]:
```suggestion
  if (_tikhonov_coeff > 0.0)
  {
    Real param_norm_sqr = 0;
    for (const auto & data : _parameters)
      for (const auto & val : *data)
        param_norm_sqr += val * val;

    val += _tikhonov_coeff * param_norm_sqr;
  }
```

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C (hunk 1 above)
Comment (zachmprince) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:97]:
```suggestion
    if (_tikhonov_coeff == 0.0)
      continue;
    auto params = _parameters[param_group_id];
    auto grads = _gradients[param_group_id];
    for (const auto & param_id : make_range(_nvalues[param_group_id]))
      grads->at(param_id) += params->at(param_id) * _tikhonov_coeff;
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Just some micro optimization. Looks good!

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C (hunk 1 above)
Comment (loganharbour) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:97]: Why? Please don't do this - bracket operator can have accidental insertion. You expect the param ID to be in the map, so write code that describes that intention (at has bounds checking).

Comment (loganharbour) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:97]: ...I say this after spending multiple days debugging... not realizing that an accidental insertion was being done silently without any error

Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: I keep forgetting about this.

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C (hunk 2 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: fixed

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C (hunk 1 above)
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: fixed

Comment (zachmprince) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: These aren't maps, they are vectors...

Comment (zachmprince) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: Bracket is more appropriate, it's not inserting anything.

Comment (loganharbour) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C]: Looks good to me. map::operator[] triggers me

**Comment by lynnmunday** [comment]
@zachmprince will you review and merge?

**Review by zachmprince — APPROVED** [review]

**Pull request merged**